README.md
# node-dev-input-reader

Read input from /dev/input/* on a linux Host.

## Feature

* Detect double press / click
* Detect long press / click
* Detect Key combinaison
* Support hardware disconnect / reconnect.
* Can work from /dev/input/by-id/ and /dev/input/by-path/
* Detect device type (event/joystick/mouse)
* Writen in Typescript

## Usage

With raw events:
```typescript
    // use RAW event
    const reader = new DevInputReader('event0');
    reader.on('error', console.error)
      .on("keyup", (data) => console.log('keyup:', data))
      .on("keypress", (data) => console.log('keypress:', data));

```

With syntetic events (recommanded):
```typescript
    // use simple events
    const reader = new DevInputReader('event0'm { retryInterval: 10000});
    reader.on('error', console.error)
      .on("key", (data) => console.log('key:', data))
      .on("error", (data) => console.log('an error occure:', data))
    // data contain the durration of the press and the concurently press key
    // in case oh hardware lost, will reconnect every retryInterval 10sec
```

With syntetic events with double/long click detection (if needed):
```typescript
    // use simple events
    const reader = new DevInputReader('event0', { retryInterval: 10000, longPress: 5000, doublePress: 300 });
    reader.on('error', console.error)
      .on("simple", (data) => console.log('simple press:', data))
      .on("double", (data) => console.log('double press:', data))
      .on("long", (data) => console.log('long press:', data))
    // long press is trigered if you keep a key pressed for more than 5 seconds
    // double press is trigered if press a key 2 time within 300 ms
    // in case oh hardware lost, will reconnect every retryInterval 10sec
```

## FAQ

> Why a new Module ?

There is a lot of existing module that can deal with input:
- [dev-input](https://www.npmjs.com/package/dev-input) Looks nice in typescript, but no docs, only used by hist owner.
- [linux-keyboard-catcher](https://www.npmjs.com/package/linux-keyboard-catcher) highly Keyboard oriented, simple interface.
- [node-keylogger](https://www.npmjs.com/package/node-keylogger) very low level + forck from [node-keyboard](https://github.com/Bornholm/node-keyboard) + contains bug..
- [raw-keyboard](https://www.npmjs.com/package/raw-keyboard) very low level + use C code.

Relative package:
- [linux-key-info](https://www.npmjs.com/package/linux-key-info) Key mapping for keyboard, I may integrate it on day.


CHANGELOG.md
# changelog


## 0.0.2 / 2020-10-05

- add missing timestamp in long events
- add CHANGELOG.md

## 0.0.1 / 2020-10-05

- initial verison


test/test.ts
import assert from 'assert';

import { delay, DevInputReader, KbEvent, UnixTimeval, AllEventsType, SimpleEventsType, SynticEventsType, KbEvent2 } from "../src/";
import KeysCodes from "../src/KeysCodes";

DevInputReader.registerKey(288, 'PAD_X');
DevInputReader.registerKey(289, 'PAD_A');
DevInputReader.registerKey(290, 'PAD_B');
DevInputReader.registerKey(291, 'PAD_Y');
DevInputReader.registerKey(292, 'PAD_L1');
DevInputReader.registerKey(294, 'PAD_R1');
DevInputReader.registerKey(296, 'PAD_SELECT');
DevInputReader.registerKey(297, 'PAD_START');
DevInputReader.registerKey(272, 'MOUSE_Clic_L');
DevInputReader.registerKey(273, 'MOUSE_Clic_R');
DevInputReader.registerKey(274, 'MOUSE_Clic_M');

/**
 * https://www.kernel.org/doc/Documentation/input/input.txt
 * https://www.kernel.org/doc/Documentation/input/joystick-api.txt
 * 
 */
/**
 * joystick-api.txt
 * https://www.mjmwired.net/kernel/Documentation/input/joystick-api.txt
 * 
 */


function startTest(node: string) {
  const K0 = new DevInputReader(node, { retryInterval: 500, longPress: 2000, doublePress: 300 });
  // const K0 = new DevInputReader(node, {retryInterval: 5000} );
  K0.on('error', console.error)
    .on('connecting', () => console.log('cnx...'))
    // .on("keypress", (data) => console.log('keypress:', JSON.stringify(data)))
    .on("long", (data) => console.log('long:', JSON.stringify(data)))
    .on("simple", (data) => console.log('simple:', JSON.stringify(data)))
    .on("double", (data) => console.log('double:', JSON.stringify(data)))
}
//startTest('event5');
//startTest('usb-0810_usb_gamepad-event-joystick');
// startTest('usb-Antecer_AmusingKeyPadK6-event-kbd');

const privateDigestEvent = (reader: DevInputReader, event: KbEvent): Promise<void> => (reader as any).digestEvent(event);

const newEvent = (type: SimpleEventsType, keyCode: number): KbEvent => {
  let now = Date.now();
  const time = new UnixTimeval(Math.floor(now / 1000), 1000 * (now % 1000));
  return {
    dev: 'dumy',
    time,
    keyCode: keyCode,
    keyName: KeysCodes[keyCode],
    type,
  } as KbEvent;
}

async function testAll() {
  const reader = new DevInputReader('dummy', { longPress: 2000, doublePress: 300 });
  const digestEvent = (event: KbEvent): Promise<void> => privateDigestEvent(reader, event);
  digestEvent(newEvent('keydown', 48));
  await delay(100);
  digestEvent(newEvent('keyup', 48));
}


function listenAction(reader: DevInputReader, actions: AllEventsType[]): string[] {
  const actionSet = new Set(actions);
  const array = [] as string[];


  const pushEv = (t: string) => (ev: KbEvent2 | KbEvent) => {
    let txt = `${t}${ev.keyName.replace(/KEY_/g, '')}`;
    if ((ev as KbEvent2).keyCodePressed && (ev as KbEvent2).keyCodePressed.length)
      txt += `[${(ev as KbEvent2).keyNamePressed.join(',').replace(/KEY_/g, '')}]`;
    array.push(txt)
  }


  if (actionSet.has("keyup"))
    reader.on("keyup", pushEv("up"));
  if (actionSet.has("keydown"))
    reader.on("keydown", pushEv("down"));
  if (actionSet.has("key"))
    reader.on("key", pushEv('key'));
  if (actionSet.has("simple"))
    reader.on("simple", pushEv('simple'));
  if (actionSet.has("double"))
    reader.on("double", pushEv('double'));
  if (actionSet.has("long"))
    reader.on("long", pushEv('long'));
  return array;
}


export enum TestKey {
  B = 48,
  N = 49,
}




describe('simple', () => {
  {
    const reader = new DevInputReader('dummy');
    const array = listenAction(reader, ['keyup', 'keydown']);
    const ev = (type: SimpleEventsType, keyCode: number) => privateDigestEvent(reader, newEvent(type, keyCode));
    it('single key', async () => {
      ev('keydown', TestKey.B);
      await delay(10);
      ev('keyup', TestKey.B);
      assert.strictEqual(array.join(','), 'downB,upB')
      await delay(10);
      assert.strictEqual(array.join(','), 'downB,upB')
    });
  }

  {
    const reader = new DevInputReader('dummy');
    const array = listenAction(reader, ['simple', 'double', 'long']);
    const ev = (type: SimpleEventsType, keyCode: number) => privateDigestEvent(reader, newEvent(type, keyCode));

    it('single key Adv', async () => {
      ev('keydown', 48);
      await delay(10);
      ev('keyup', TestKey.B);
      await delay(10);
      assert.strictEqual(array.join(','), 'simpleB')
      await delay(10);
    });
    array.splice(0, array.length)

    it('single key Adv', async () => {
      ev('keydown', TestKey.B);
      await delay(100);
      ev('keydown', TestKey.N);
      await delay(100);
      ev('keyup', TestKey.B);
      await delay(100);
      ev('keyup', TestKey.N);
      await delay(100);
      assert.strictEqual(array.join(','), 'simpleB,simpleB[N],simpleN')
      await delay(10);
    });

  }

});

// private async digestEvent(event: KbEvent): Promise<void> {
// const K6 = new LinuxInputEvent('usb-Antecer_AmusingKeyPadK6-event-kbd');
// Ok 
// K6.on("keypress", (data) => { console.log(data) });
// Ok 
// K6.on("key", (data) => { console.log(data) });
// K6.on("double", (data) => { console.log(data) });
// K6.on("simple", (data) => { });
// const K5 = new Keyboard('event5');


test/testRaw.ts
import assert from 'assert';
import * as DIR from "../src";
import { TestKey, preparEnv } from "./commonTest";

describe('Raw Test', () => {
  {
    const { array, ev } = preparEnv(['keyup', 'keydown']);
    it('single down/up', async () => {
      ev('keydown', TestKey.B);
      await DIR.delay(1);
      ev('keyup', TestKey.B);
      assert.strictEqual(array.join(','), 'downB,upB')
    });

    it('no phantom key event', async () => {
      await DIR.delay(1);
      assert.strictEqual(array.join(','), 'downB,upB')
    });

    it('mutiples down/up', async () => {
      array.splice(0, array.length)
      ev('keydown', TestKey.B);
      await DIR.delay(1);
      ev('keyup', TestKey.B);
      await DIR.delay(1);
      ev('keydown', TestKey.N);
      await DIR.delay(1);
      ev('keyup', TestKey.N);
      await DIR.delay(1);
      ev('keydown', TestKey.N);
      await DIR.delay(1);
      ev('keyup', TestKey.N);
      await DIR.delay(1);
      assert.strictEqual(array.join(','), 'downB,upB,downN,upN,downN,upN')
    });

    it('mutiples down/up concurent key', async () => {
      array.splice(0, array.length)
      ev('keydown', TestKey.B);
      await DIR.delay(1);
      ev('keydown', TestKey.N);
      await DIR.delay(1);
      ev('keyup', TestKey.B);
      await DIR.delay(1);
      ev('keyup', TestKey.N);
      await DIR.delay(1);
      assert.strictEqual(array.join(','), 'downB,downN,upB,upN')
    });
  }
});


test/testLong.ts
import assert from 'assert';
import * as DIR from "../src";
import { TestKey, preparEnv } from "./commonTest";

describe('sintetic event + long detection:', () => {
    {
        const { array, ev } = preparEnv(['simple', 'double', 'long'], { longPress: 5 });
        it('waiting longpress', async () => {
            array.splice(0, array.length);
            ev('keydown', 48);
            await DIR.delay(1);
            assert.strictEqual(array.join(','), '')
        });

        it('longpress', async () => {
            await DIR.delay(5);
            assert.strictEqual(array.join(','), 'longB[B]')
        });

        it('no extra event on up', async () => {
            ev('keyup', TestKey.B);
            await DIR.delay(1);
            assert.strictEqual(array.join(','), 'longB[B]')
        });

        it('simple on longpress', async () => {
            array.splice(0, array.length);
            ev('keydown', 48);
            await DIR.delay(1);
            ev('keyup', 48);
            assert.strictEqual(array.join(','), 'simpleB')
        });
    }
});


src/KeysCodes.ts
const KeysCodes: { [key: number]: string } = {
    1: 'KEY_ESC',
    2: 'KEY_1',
    3: 'KEY_2',
    4: 'KEY_3',
    5: 'KEY_4',
    6: 'KEY_5',
    7: 'KEY_6',
    8: 'KEY_7',
    9: 'KEY_8',
    10: 'KEY_9',
    11: 'KEY_0',
    12: 'KEY_MINUS',
    13: 'KEY_EQUAL',
    14: 'KEY_BACKSPACE',
    15: 'KEY_TAB',
    16: 'KEY_Q',
    17: 'KEY_W',
    18: 'KEY_E',
    19: 'KEY_R',
    20: 'KEY_T',
    21: 'KEY_Y',
    22: 'KEY_U',
    23: 'KEY_I',
    24: 'KEY_O',
    25: 'KEY_P',
    26: 'KEY_LEFTBRACE',
    27: 'KEY_RIGHTBRACE',
    28: 'KEY_ENTER',
    29: 'KEY_LEFTCTRL',
    30: 'KEY_A',
    31: 'KEY_S',
    32: 'KEY_D',
    33: 'KEY_F',
    34: 'KEY_G',
    35: 'KEY_H',
    36: 'KEY_J',
    37: 'KEY_K',
    38: 'KEY_L',
    39: 'KEY_SEMICOLON',
    40: 'KEY_APOSTROPHE',
    41: 'KEY_GRAVE',
    42: 'KEY_LEFTSHIFT',
    43: 'KEY_BACKSLASH',
    44: 'KEY_Z',
    45: 'KEY_X',
    46: 'KEY_C',
    47: 'KEY_V',
    48: 'KEY_B',
    49: 'KEY_N',
    50: 'KEY_M',
    51: 'KEY_COMMA',
    52: 'KEY_DOT',
    53: 'KEY_SLASH',
    54: 'KEY_RIGHTSHIFT',
    55: 'KEY_KPASTERISK',
    56: 'KEY_LEFTALT',
    57: 'KEY_SPACE',
    58: 'KEY_CAPSLOCK',
    59: 'KEY_F1',
    60: 'KEY_F2',
    61: 'KEY_F3',
    62: 'KEY_F4',
    63: 'KEY_F5',
    64: 'KEY_F6',
    65: 'KEY_F7',
    66: 'KEY_F8',
    67: 'KEY_F9',
    68: 'KEY_F10',
    69: 'KEY_NUMLOCK',
    70: 'KEY_SCROLLLOCK',
    71: 'KEY_KP7',
    72: 'KEY_KP8',
    73: 'KEY_KP9',
    74: 'KEY_KPMINUS',
    75: 'KEY_KP4',
    76: 'KEY_KP5',
    77: 'KEY_KP6',
    78: 'KEY_KPPLUS',
    79: 'KEY_KP1',
    80: 'KEY_KP2',
    81: 'KEY_KP3',
    82: 'KEY_KP0',
    83: 'KEY_KPDOT',
    85: 'KEY_ZENKAKUHANKAKU',
    86: 'KEY_102ND',
    87: 'KEY_F11',
    88: 'KEY_F12',
    89: 'KEY_RO',
    90: 'KEY_KATAKANA',
    91: 'KEY_HIRAGANA',
    92: 'KEY_HENKAN',
    93: 'KEY_KATAKANAHIRAGANA',
    94: 'KEY_MUHENKAN',
    95: 'KEY_KPJPCOMMA',
    96: 'KEY_KPENTER',
    97: 'KEY_RIGHTCTRL',
    98: 'KEY_KPSLASH',
    99: 'KEY_SYSRQ',
    100: 'KEY_RIGHTALT',
    102: 'KEY_HOME',
    103: 'KEY_UP',
    104: 'KEY_PAGEUP',
    105: 'KEY_LEFT',
    106: 'KEY_RIGHT',
    107: 'KEY_END',
    108: 'KEY_DOWN',
    109: 'KEY_PAGEDOWN',
    110: 'KEY_INSERT',
    111: 'KEY_DELETE',
    113: 'KEY_MUTE',
    114: 'KEY_VOLUMEDOWN',
    115: 'KEY_VOLUMEUP',
    116: 'KEY_POWER',
    117: 'KEY_KPEQUAL',
    119: 'KEY_PAUSE',
    121: 'KEY_KPCOMMA',
    122: 'KEY_HANGUEL',
    123: 'KEY_HANJA',
    124: 'KEY_YEN',
    125: 'KEY_LEFTMETA',
    126: 'KEY_RIGHTMETA',
    127: 'KEY_COMPOSE',
    128: 'KEY_STOP',
    129: 'KEY_AGAIN',
    130: 'KEY_PROPS',
    131: 'KEY_UNDO',
    132: 'KEY_FRONT',
    133: 'KEY_COPY',
    134: 'KEY_OPEN',
    135: 'KEY_PASTE',
    136: 'KEY_FIND',
    137: 'KEY_CUT',
    138: 'KEY_HELP',
    183: 'KEY_F13',
    184: 'KEY_F14',
    185: 'KEY_F15',
    186: 'KEY_F16',
    187: 'KEY_F17',
    188: 'KEY_F18',
    189: 'KEY_F19',
    190: 'KEY_F20',
    191: 'KEY_F21',
    192: 'KEY_F22',
    193: 'KEY_F23',
    194: 'KEY_F24',
    240: 'KEY_UNKNOWN'
  }
  
  export default KeysCodes;

src/InputInterfaces.ts
import { EVENT_TYPES } from "./enums";
import UnixTimeval from "./UnixTimeval";

export interface JsEvent {
    dev: string;
    ts: number,
    value: number;
    /**
     * #define JS_EVENT_BUTTON         0x01 // button pressed/released
     * #define JS_EVENT_AXIS           0x02 // joystick moved 
     * #define JS_EVENT_INIT           0x80 // initial state of device
     */
    type: number;
    /**
     * 1st Axis X	0
     * 1st Axis Y	1
     * 2nd Axis X	2
     * 2nd Axis Y	3
     */
    number: number,
}

export interface KbEvent {
    readonly dev: string;
    readonly time: UnixTimeval,
    readonly keyCode: number;
    readonly keyName: string,
    readonly type: typeof EVENT_TYPES[number],
}

export interface KbEvent2 extends KbEvent {
    readonly keyCodePressed: number[];
    readonly keyNamePressed: string[];
    readonly durationMs: number;
}


src/DevInputReader.ts
import fs, { ReadStream } from 'fs';
import { EventEmitter } from 'events';
import KeysCodes from './KeysCodes';
import delay from './delay';
import { KbEvent2, JsEvent, KbEvent } from './InputInterfaces';
import UnixTimeval from './UnixTimeval';
import { SUPPORTED_USB_DEVICE_TYPES, SUPPORTED_USB_DEVICE_TYPES_SET, EVENT_TYPES } from './enums';
import KeyInternalState, { KeyState } from './KeyInternalState';

export const standardEvents = ['error', 'connecting'] as const;
export const simpleEvents = ['keyup', 'keypress', 'keydown'] as const; // Keep this order
export const synticEvents = ['key', 'long', 'double', 'simple'] as const;

export type StandardEventsType = typeof standardEvents[number];
export type SimpleEventsType = typeof simpleEvents[number];
export type SynticEventsType = typeof synticEvents[number];

export type AllEventsType = StandardEventsType | SimpleEventsType | SynticEventsType | 'joystick';

export const specialEventsSet = new Set([...simpleEvents, ...synticEvents]) as Set<string>;

export interface DevInputReaderOption {
    retryInterval?: number;
    doublePress?: number;
    longPress?: number;
}

export default class DevInputReader extends EventEmitter {
    dev: string;
    devType?: typeof SUPPORTED_USB_DEVICE_TYPES[number];
    // eventPath?: string;
    data: ReadStream | null = null;
    // public keys = toKey;
    private options: DevInputReaderOption;

    listener = 0;

    public static registerKey(id: number, name: string): void {
        KeysCodes[id] = name;
    }

    constructor(dev: string, options?: DevInputReaderOption) {
        super();
        options = options || {};
        this.options = { ...options };
        this.dev = dev || 'event0';
    }

    private async startLoop() {
        while (this.options.retryInterval) {
            this.data = null;
            if (this.start()) {
                return;
            }
            await delay(this.options.retryInterval);
        }
    }

    private start() {
        this.emit('connecting', '');
        let eventPath = '';
        // this.buffer = Buffer.alloc(this.bufferSize);
        if (this.dev === 'dummy') {
            // for test only
            this.devType = 'event';
            return true;
        } else if (this.dev.startsWith('event'))
            eventPath = `/dev/input/${this.dev}`;
        else if (this.dev.startsWith('pci-'))
            eventPath = `/dev/input/by-path/${this.dev}`;
        else if (this.dev.startsWith('latform-'))
            eventPath = `/dev/input/by-path/${this.dev}`;
        else if (this.dev.startsWith('/'))
            eventPath = this.dev;
        else
            eventPath = `/dev/input/by-id/${this.dev}`;
        let lnk = eventPath;
        try {
            if (eventPath)
                fs.statSync(eventPath);
        } catch (e) {
            this.emit('error', Error(`Can not find dev "${eventPath}"`));
            return false;
        }
        try {
            if (eventPath)
                lnk = fs.readlinkSync(eventPath, 'utf8');
        } catch (e) {
            // keep lnk as its
        }
        const m = lnk.match(/\/([a-z]+)\d+$/);
        if (!m) {
            this.emit('error', Error(`Can Not identify device type from: "${lnk}"`));
            return false;
        }
        if (!SUPPORTED_USB_DEVICE_TYPES_SET.has(m[1])) {
            this.emit('error', Error(`Unknow input type for: ${lnk} Supported:${SUPPORTED_USB_DEVICE_TYPES.join(', ')}`));
            return false;
        }
        this.devType = m[1] as typeof SUPPORTED_USB_DEVICE_TYPES[number];

        if (!this.data) {
            this.data = fs.createReadStream(eventPath);
            if (this.devType === 'event')
                this.data.on('data', (data: Buffer) => {
                    const event: KbEvent | null = this.parseEvent(data);
                    if (event) {
                        this.digestEvent(event);
                    }
                });
            else if (this.devType === 'js')
                this.data.on('data', (data: Buffer) => {
                    const event: JsEvent | null = this.parseJoystick(data);
                    if (event) {
                        this.emit('joystick', event)
                    }
                });
            else if (this.devType === 'mouse')
                this.data.on('data', (data: Buffer) => {
                    this.parseMouse(data);
                });

            this.data.on('error', async err => {
                this.emit('error', err);
                if (this.options.retryInterval) {
                    await delay(this.options.retryInterval);
                    this.data = null;
                    this.startLoop();
                }
            });
        }
        return true;
    }

    private stop() {
        if (this.data) {
            this.data.close();
            this.data = null;
        }
    }

    private keyStatus: { [key: string]: KeyInternalState } = {};
    private keyCodePresses: Set<number> = new Set();
    private keyNamePresses: Set<string> = new Set();

    private async digestEvent(event: KbEvent): Promise<void> {
        const type: SimpleEventsType = event.type;

        this.emit(type, event);
        // get / init key internal state
        let keyStatus = this.keyStatus[event.keyCode];
        if (!keyStatus) {
            keyStatus = new KeyInternalState();
            this.keyStatus[event.keyCode] = keyStatus;
        }
        switch (event.type) {
            case 'keydown':
            case 'keypress': {
                keyStatus.state = KeyState.Down;
                keyStatus.prevDown = keyStatus.down;
                keyStatus.down = event.time;
                this.keyCodePresses.add(event.keyCode);
                if (event.keyName)
                    this.keyNamePresses.add(event.keyName);

                if (this.options.longPress) {
                    await delay(this.options.longPress);

                    if (keyStatus.state === KeyState.Down && keyStatus.down === event.time) {
                        this.emit('long', {
                            dev: this.dev,
                            time: keyStatus.down,
                            durationMs: this.options.longPress,
                            keyCode: event.keyCode,
                            keyName: event.keyName,
                            type: event.type,
                            keyCodePressed: [...this.keyCodePresses],
                            keyNamePressed: [...this.keyNamePresses],
                        } as KbEvent2);
                        keyStatus.flush();
                    }
                }
                break;
            }
            case 'keyup': {
                this.keyCodePresses.delete(event.keyCode);
                if (event.keyName)
                    this.keyNamePresses.delete(event.keyName);
                if (!keyStatus.down) {
                    // a long press already occured
                    return;
                }
                keyStatus.up = event.time;
                if (keyStatus.state === KeyState.Down) {
                    this.emit('key', {
                        dev: this.dev,
                        time: keyStatus.up,
                        durationMs: keyStatus.up.fromMs(keyStatus.down),
                        keyCode: event.keyCode,
                        keyName: event.keyName,
                        type: event.type,
                    } as KbEvent2);
                }
                keyStatus.state = KeyState.Up;
                keyStatus.prevUp = keyStatus.up;
                const durationDoubleMs = keyStatus.up.fromMs(keyStatus.prevDown);
                if (this.options.doublePress && (durationDoubleMs < this.options.doublePress)) {
                    this.emit('double', {
                        dev: this.dev,
                        time: keyStatus.up,
                        durationMs: durationDoubleMs,
                        keyCode: event.keyCode,
                        keyName: event.keyName,
                        type: event.type,
                        keyCodePressed: [...this.keyCodePresses],
                        keyNamePressed: [...this.keyNamePresses],
                    } as KbEvent2);
                    keyStatus.flush();
                } else {
                    if (this.options.doublePress)
                        await delay(this.options.doublePress)
                    if (keyStatus.up === event.time) {
                        this.emit('simple', {
                            dev: this.dev,
                            time: keyStatus.up,
                            durationMs: keyStatus.up.fromMs(keyStatus.down),
                            keyCode: event.keyCode,
                            keyName: event.keyName,
                            type: event.type,
                            keyCodePressed: [...this.keyCodePresses],
                            keyNamePressed: [...this.keyNamePresses],
                        } as KbEvent2);
                    }
                }
                break;
            }
        }
    }

    private parseEvent(data: Buffer): KbEvent | null {
        // if (this.devType === 'event') {
        if (data.length >= 48) {
            // 72 or 96
            const offset = 24;
            const evType = data.readUInt16LE(offset + 16);
            if (evType === 1) { // const EV_KEY = 1;
                const tv_sec = data.readUInt32LE(offset + 0);
                const tv_usec = data.readUInt32LE(offset + 8);
                const keyCode = data.readUInt16LE(offset + 18);
                const type = data.readUInt32LE(offset + 20);
                return {
                    dev: this.dev,
                    time: new UnixTimeval(tv_sec, tv_usec),
                    keyCode,
                    keyName: KeysCodes[keyCode],
                    type: EVENT_TYPES[type],
                } as KbEvent;
            }
        }
        return null;
    }

    private parseJoystick(data: Buffer): JsEvent | null {
        // } else if (this.devType === 'js') {
        const ts = data.readUInt32LE(0);
        const value = data.readInt16LE(4);
        const type = data.readUInt8(6);
        const number = data.readUInt8(7);
        return {
            dev: this.devType,
            ts,
            value,
            type,
            number,
        } as JsEvent
        // this.emit('joystick',  as JsEvent)
        // return null;
    }

    private parseMouse(data: Buffer): KbEvent | null {
        // } else if (this.devType === 'mouse') {
        // legacy mouse sould not be use anymore.
        if (data.length === 3) {
            const v1 = data.readUInt8(0);
            const v2 = data.readUInt8(1);
            const v3 = data.readUInt8(2);
            console.log(`mouse ${data.length}: ${v1} ${v2} ${v3}`);
        }
        //} else {
        //    console.log(`missing implementation for ${this.devType}`);
        //}
        return null;
    }

    ///////////////////////////
    // @param event EVENTS
    //
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    once(event: string, listener: (...args: any[]) => void): this {
        if (specialEventsSet.has(event))
            throw Error(`once not implemented on event: "${event}"`);
        return super.once(event, listener);
    }

    removeAllListeners(event: string): this {
        if (specialEventsSet.has(event))
            throw Error(`removeAllListeners not implemented on event: "${event}"`);
        return super.removeAllListeners(event);
    }

    on(event: SimpleEventsType, listener: (event: KbEvent) => void): this;
    on(event: SynticEventsType, listener: (event: KbEvent2) => void): this;
    on(event: 'joystick', listener: (event: JsEvent) => void): this;
    on(event: 'connecting', listener: (event: string) => void): this;
    on(event: 'error', listener: (event: Error) => void): this;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    on(event: AllEventsType, listener: (...args: any[]) => void): this {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return this.addListener(event as any, listener);
    }

    addListener(event: SimpleEventsType, listener: (event: KbEvent) => void): this;
    addListener(event: SynticEventsType, listener: (event: KbEvent2) => void): this;
    addListener(event: 'joystick', listener: (event: JsEvent) => void): this;
    addListener(event: 'connecting', listener: (event: string) => void): this;
    addListener(event: 'error', listener: (event: Error) => void): this;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    addListener(event: AllEventsType, listener: (...args: any[]) => void): this {
        super.on(event, listener);
        if (specialEventsSet.has(event) && ++this.listener === 1) {
            this.startLoop();
        }
        return this;
    }

    off(event: SimpleEventsType, listener: (event: KbEvent) => void): this;
    off(event: SynticEventsType, listener: (event: KbEvent2) => void): this;
    off(event: 'joystick', listener: (event: JsEvent) => void): this;
    off(event: 'connecting', listener: (event: string) => void): this;
    off(event: 'error', listener: (error: Error) => void): this;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    off(event: AllEventsType, listener: (...args: any[]) => void): this {
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        return this.removeListener(event as any, listener);
    }

    removeListener(event: SimpleEventsType, listener: (event: KbEvent) => void): this;
    removeListener(event: SynticEventsType, listener: (event: KbEvent2) => void): this;
    removeListener(event: 'joystick', listener: (event: JsEvent) => void): this;
    removeListener(event: 'connecting', listener: (event: string) => void): this;
    removeListener(event: 'error', listener: (error: Error) => void): this;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    removeListener(event: AllEventsType, listener: (...args: any[]) => void): this {
        super.off(event, listener);
        if (specialEventsSet.has(event) && --this.listener === 0) {
            this.stop();
        }
        return this;
    }

    emit(event: SimpleEventsType, data: KbEvent): boolean;
    emit(event: SynticEventsType, data: KbEvent2): boolean;
    emit(event: 'joystick', data: JsEvent): boolean;
    emit(event: 'connecting', data: string): boolean;
    emit(event: 'error', data: Error): boolean;

    // eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/no-explicit-any
    emit(event: AllEventsType, data: any): boolean {
        return super.emit(event, data);
    }
}
